### ISSUE TITLE
BlenderBIM: visualise layered walls from IfcMaterialLayerSets?

### ISSUE THREAD

**Issue by theoryshaw** [IfcOpenShell/IfcOpenShell#1227]
Would be awesome to hardcode a Sverchok script like [this](https://gitlab.com/osarch/Example_Files/-/blob/ef018c39abe1de9364c62a73f0a9999ade77cf3b/Blender/Sverchok/Wall_Multiple_Layers_Offset_Extrusion.blend) to create multi-layer walls from walls that have IfcMaterialLayerSets?  
Background discussion [here](https://community.osarch.org/discussion/389/sverchok-offset-one-side).

![3m4iprf6aof9](https://user-images.githubusercontent.com/507113/103485675-c093c700-4dbd-11eb-8e3d-913a855da501.gif)

**Comment by Moult** [comment]
Absolutely correct - there needs to be support for visualising material layers. There are a variety of approaches to be considered, each with pros and cons, including:

 1. Sverchok, as you've shown
 2. Geometry nodes, coming soon
 3. bgl decoration
 4. Material shaders
 5. Actual geometry generation (e.g. the IfcConvert method)

**Comment by aothms** [comment]
> Actual geometry generation (e.g. the IfcConvert method)

Note that this is not exclusive to IfcConvert. (`ifcopenshell.geom.settings(APPLY_LAYERSETS=True)`)

To do this robustly for curved walls, maybe inside IfcOpenShell will work best?

**Comment by theoryshaw** [comment]
@Moult , @aothms 
Now that this has percolated a little, what approach do you guys think would be best here?
I'm thinking of putting together another Bountysource campaign for this functionality as well. 
If so, it would be nice to outline a roadmap of suggested best approach--something to point to, that better defines the task and makes it more tangle. 
Perhaps start another issue tracker?

**Comment by Moult** [comment]
@theoryshaw sorry recently I've been a little distracted in the sequencing and costing domain due to an application within the company I work for.

I think @aothms 's suggestion should just work. Perhaps wait a bit until I get back to regular authoring + drawings and I'll try out the suggestion and see what that gets us.

**Comment by Moult** [comment]
@aothms I tried apply layersets with this file: https://github.com/ENAC-CNPA/BIMxBEM/blob/master/IfcTestFiles/2Storey_2x3_A22.ifc but did not see any difference. What am I missing?

**Comment by theoryshaw** [comment]
With the [current paradigm](https://github.com/IfcOpenShell/IfcOpenShell/commit/5c1d26ac91f88b9134c714f2cba2d4987c9f30f1) of encoding the overall floor/wall's thickness into a blender solidifier modifier, in the future roadmap, how do you see codifying and visualizing multiple IfcMaterialLayerSets layers in Blender?

**Comment by Moult** [comment]
The floor is stored in a solidify modifier, but a wall is not. This is on purpose to allow for mitred joints.

I have no idea how to do it yet :) One option I want to explore is the apply layersets but I haven't gotten the "hello world" working yet. But that's OK, I'm sure over time a solution will present itself, if not to me, to someone else hacking on the code!

Layer sets can get quite complex I believe including priorities, but perhaps for 99% of the time we can get away with a simpler heuristic, maybe drawn via a shader, something via bgl, or generated case by case if you're focusing on a junction.

**Comment by theoryshaw** [comment]
Hi @aothms, what's your thoughts here?  Would love to see layered walls/floors/roofs, etc... would be a big step for BlenderBIM, i think.

**Comment by aothms** [comment]
> @aothms I tried apply layersets with this file: https://github.com/ENAC-CNPA/BIMxBEM/blob/master/IfcTestFiles/2Storey_2x3_A22.ifc but did not see any difference. What am I missing?

Not entirely sure. I don't have a lot of confidence in the layerset slicing code. I removed one condition that I didn't understand anymore and now it seems to work.

![afbeelding](https://user-images.githubusercontent.com/1096535/133968805-5f263c44-375a-462e-b0a8-8ed2e8ca9657.png)

I think for an authoring environment more might be needed though in terms of an API? 

Also in general there is the trick that for performance we should do this in 2D. So footprint -> slice -> extrude. Instead of footprint -> extrude -> slice.

**Comment by EAzari** [comment]
The preferred approach should be like this:
1. You design each layer and its relevant (parametric) geometry separately, for example, a wall with these layers:
  - Exterior finish 
  - Exterior insulation(s)
  - Exterior frame(s)
  - Core
  - Interior frame(s)
  - Interior insulation(s)
  - Interior finish
  2. Join or Attach layers together to have a whole-part
  3. Make it a family or an (reusable) object --> You choose the family type and that would become default one two draw walls that have those characteristics 
  
  I mean you think about each layer, built each one, and put all together as a whole, and it becomes your family or whatever you want to name it

**Comment by theoryshaw** [comment]
related conversation: https://community.osarch.org/discussion/comment/9592

**Comment by theoryshaw** [comment]
found in the wild: http://blog.ifcopenshell.org/2015/10/separating-elements-by-their-material.html

**Comment by theoryshaw** [comment]
also related
- https://forum.freecadweb.org/viewtopic.php?t=15080
- https://sourceforge.net/p/ifcopenshell/discussion/1782718/thread/38f13717/

**Comment by Moult** [comment]
@aothms revisiting this, I've applied layerset slicing and the slice now works for a simple wall. However in the SVG there seems to be no indication of the material of the layer. Inside the group of the product, I see three subpaths, one for each layer, but each path does not have any metadata. Any ideas how to tackle this?

![2023-03-06-153754_1057x654_scrot](https://user-images.githubusercontent.com/88302/223020802-3f250e98-a510-4129-babc-a06edf98f732.png)

**Comment by aothms** [comment]
Hm, good one. Unfortunately we don't know a lot (this iterator was for quick and dirty 3d viz after all) but I think I can sneak in the surface style numeric id in there somehow. If you have a one to one mapping in python between surface style and material it might be enough?

**Comment by Moult** [comment]
Unfortunately no there is no guarantee of a 1:1 mapping between surface styles and materials :(

**Comment by Moult** [comment]
If IFC were completely used for drawing styling, I'd say that it's a step in the right direction, for example surface styles for projection hatching and fill area styles for cut hatching (unofficially, just me being imaginative).

However, I'm increasingly of the opinion that IFC shouldn't be the one reinventing the whole visual style game when there is CSS and SVG already out there for the 2D world. Let IFC do what it does best: semantics of the built environment, and let other standards do what they also do best. In a "drawingless future" I see the semantics of the drawings (or maybe "diagrams" as they will transition towards) as being more important. So, given that the material layer set is made out of layered materials, I would personally prefer materials as an ideal SVG output.

**Comment by Moult** [comment]
Bump? :)

**Comment by Moult** [comment]
https://github.com/IfcOpenShell/IfcOpenShell/pull/2859 just reminded me that this is another reason why it's important to denote the material, as material sets affect weight calculations, of which the density is stored on materials.

**Comment by Moult** [comment]
Bump @aothms :)

**Comment by Moult** [comment]
Here's a small test with apply_layersets with a few junctions with the relevant relconnectspath relationships. I'm noticing no impact at all of the rel connects on the layers (i.e. they don't turn the corner). Is there something I'm missing apart from `settings.set(settings.APPLY_LAYERSETS, True)`?

[junctions.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/11322904/junctions.txt)


![2023-04-25-232036_1020x800_scrot](https://user-images.githubusercontent.com/88302/234289365-550e1e76-e250-4544-bf80-0e51ca0096b9.png)

**Comment by Moult** [comment]
I had a shot at implementing this in pure Python with a few assumptions (e.g. does not handle curved walls yet) and I think this produces correct results in the viewport. It also seems pretty fast.

Note that this is _only_ in the viewport. In theory the code can be reused (with a bit of extra work to create closed polygons) for drawing generation so we'll see how this goes. In the future I would prefer to delete this code and use APPLY_LAYERSETS instead but at least we have something running.

![out](https://user-images.githubusercontent.com/88302/234570534-59f6b24b-2905-47e7-a496-d6668bb2544b.gif)

**Comment by aothms** [comment]
> Here's a small test with apply_layersets with a few junctions with the relevant relconnectspath relationships. I'm noticing no impact at all of the rel connects on the layers (i.e. they don't turn the corner). Is there something I'm missing apart from settings.set(settings.APPLY_LAYERSETS, True)?

Your axis representations are different from what the other tools do:

![afbeelding](https://user-images.githubusercontent.com/1096535/234583475-2463227f-013b-453e-9826-f252d4a31630.png)

In other tools the only case when you have intersecting/touching axes is when you have "diagonally" cut joins in which you do not need to fold.

![afbeelding](https://user-images.githubusercontent.com/1096535/234585296-c72c959b-b0a3-4426-8da0-d853e321badf.png)

^ You can see in the duplex model which has both kind of joins

![afbeelding](https://user-images.githubusercontent.com/1096535/234585420-5d2c420d-417e-43d0-a985-37c0ead7af7d.png)

![afbeelding](https://user-images.githubusercontent.com/1096535/234585485-06deae83-05c3-4aa0-afd4-310f0253f467.png)

I'm not saying there is a right or wrong, just explaining how you triggered the situation that according to the current algo, no folds are necessary.

Here an image of how the axis extends beyond the body:

![afbeelding](https://user-images.githubusercontent.com/1096535/234586511-6fe46dde-53b1-4ed0-86ff-dd18917913e2.png)

That is what I can see at least now. Not sure if it's the only thing.

**Comment by Moult** [comment]
Thanks! That sounds like a likely reason. Our axes extend because the axes extends to meet the joining axis. This also seems to better capture the design intent of a wall in graph view (i.e. on paper if you were to draw walls via a single line) (or beam, column, etc) extending to another wall ['s axis] in GRAPH_VIEW. The body is naturally clipped by any connects relationship.

This seems to be one of the less mature parts of the IFC spec (especially noticing the axis that stops with the door opening, which I'd say is definitely wrong). Looking at the [IfcWall docs](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcWall.htm) the first sentence suggests that the Axis is only used to locate the layers via the usage parameters (i.e. offset and direction sense) and nothing else (i.e. does not impact rel connects path):

> The 'Axis' shape representation is only used to locate the material layer set along the axis, if the [IfcMaterialLayerSetUsage](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcMaterialLayerSetUsage.htm) is applied to the [IfcWall](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcWall.htm). In this case, the wall axis is used to apply the material layer set usage parameter to the wall geometry.

That said, the diagrams in the docs offer a few more hints about how Axis is supposedly meant to be implemented, which is what has led to my implementation.

In the diagrams on the IfcWall page, the Axis is represented using a red line. That same line is shown in multiple other examples where walls join. That Axis follows the same behaviour exhibited in the BlenderBIM Add-on, that is, it extends past the body and is trimmed at the connecting wall's axis.

The same behaviour (i.e. matching the BlenderBIM Add-on's implementation) is shown in the IfcBeam's documentation: https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcBeam.htm in this case the Axis is shown as a thick black line (labeled as the axis at the cardinal point top center) which also extends to meet other axes to join, and the body is clipped (possibly by an assumed join relationship?)

Again, same behaviour for IfcColumn: https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcColumn.htm

The docs for rel connects https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcRelConnectsPathElements.htm seem to emphasize this behaviour. Less clear is the dashed black line in "Figure 6.1.3.28.C — Path connection L-Type" which may also be an Axis, which also matches the BlenderBIM Add-on's behaviour.

**Comment by aothms** [comment]
I think you're right.

One advantage is also that it makes atstart/atpath/atend less ambiguous. Because it should directly correspond to the vertices of the axes coinciding (with some tolerance).

One disadvantage is that axis length doesn't really correspond to wall length anymore. I would find topology more important though.

Why do other tools unanimously decide on disjoint wall axes though? Are we missing sth?

**Comment by theoryshaw** [comment]
If anyone wants to help fund the development of visualizing layers in BlenderBIM. 
https://opencollective.com/osarch/projects/visualize_layers_in_blenderbim

**Comment by theoryshaw** [comment]
To date, funding is currently at $1,844.36 USD
https://opencollective.com/osarch/projects/visualize_layers_in_blenderbim

**Comment by aothms** [comment]
So @theoryshaw previously had a quick chat about it. In my opinion the following needs to be done:

- C++ Efficient visualization interface to layerset slicing, so not full 3d solids as meshes, but quicker
- C++ Respect layer priorities
- C++ Allow blenderbim's connected axis representations (determine whether folds are necessary based on body, not based on axis intersection).
- Py UI to represent in bbim

![unnamed](https://github.com/IfcOpenShell/IfcOpenShell/assets/1096535/bd085a50-eb7e-4132-b175-dbef107c02cf)

What kind of interface do you imagine. I think that slicing all solids generates quite a bit of overhead (but it could be worth it as a setting). What we could also do is detect inner and outer face of the box and style accordingly. And have an overlay in the viewport on the selected object.

**Comment by theoryshaw** [comment]
Wanted to introduce [Ryan Hughes](https://community.osarch.org/discussion/comment/19499/#Comment_19499) (@rhughes42).  I connected with him, to gauge his interest in taking on this scope.   He seemed interested. :)

Hi Ryan. :)

Let us know if you have any questions. 

 > What kind of interface do you imagine. 

As i understand it, currently layers are visualized using a [shader decorator](https://github.com/IfcOpenShell/IfcOpenShell/blob/51640ccc28bac124176091d1d2f3c5d694013dff/src/blenderbim/blenderbim/bim/module/drawing/decoration.py#L1893).  Does it make sense to extend this logic?  Currently it only works in plan view, would be nice to get to work in the other context views as well. 

> I think that slicing all solids generates quite a bit of overhead (but it could be worth it as a setting)

Defer to those more knowledge on what the the exact overhead would be here.  I would only say, it would be ideal to have layers as real geometry in the scene--so one can snap to this geometry as they design.   It would also allow applying different materials to the different layer geometries.  This helps when designing in the scene to get a WYSIWYG feedback.   Maybe all this can be accomplished with decoration shaders, not individual mesh layers... not sure.

**Comment by senyai** [comment]
I hope it will be possible to add ventilated cavity walls, i.e. air between walls. There's a site ([is Russian](https://smartcalc.ru/thermocalc?&gp=210&rt=0&ct=0&os=0&ti=20&to=-10&hi=55&ho=85&ld0=500&le0=1&lt0=0&mm0=587&ld1=5000&le1=1&lt1=0&mm1=189&ld2=100&le2=1&lt2=0&mm2=151)) that allows to edit wall materials. Their UI is very simple, but powerful. 
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/76137/0f25dbcb-4e03-48b7-a5ec-da501fac70f9)
Just my 2 cents

**Comment by rhughes42** [comment]
Hey guys, thank you @theoryshaw for adding me here. I would be interested indeed to take the project. I just want to test a toy version of the problem first to get fully up to speed on all of the components at play before committing. I have two or three specific questions that I think @aothms could probably answer for me on a short call if you're willing to do that? I understand the problem, and I can see the potential issues with overhead, and the potential benefits of being able to snap to interior vertices in 3D etc. - my concern is seperating the bounding volume snappable geometry for each layer and the shader / display of the interior, for example if it needs to be hatched or dotted.

I plan to get more and more involved (I'm also starting to use BlenderBim in some other projects) and I have another project that uses a lot of computer vision that could potentially interestingly benefit from solving this problem, but in reverse. So a short session to get the lay of the land should be a good investment for the project.

@senyai something like that UI looks perfectly sufficient, at least for v1. Perhaps v2 could add patterns, and convert each list item into a autocompleting search box that then switches back to appear as-is unless selected. Just an idea. But for the first version, to make sure the core problem is solved, it looks fine.

Looking forward to hearing back. 🤘

**Comment by Moult** [comment]
@senyai air is supported as a layer in IFC :) That UI is pretty similar to how things work in IFC. It would be awesome to have a meeting to discuss how to progress this. The end usecases are:

 - Visualisation as 2D, both in the viewport as well as in SVG plan / section drawings, including handling complex scenarios such as sloped cuts.
 - Getting key points so that we can handle snapping scenarios. Probably comes for free with the 2D visualisation.
 - 3D slicing, slower, and not entirely sure when this would be used.

**Comment by aothms** [comment]
Jumping on a call sounds good to me, but I think it'd be good if @Moult also joins, because my understanding of the end goal in terms of user interface is rather minimal.

**Comment by rhughes42** [comment]
@Moult and @aothms that sounds totally fine with me. Let me grab my calendar, and you guys could see if you can find a time that works. Here's EU: https://calendar.app.google/YwPiZv3bW6BwJgXC9 and CST: https://calendar.app.google/QcQQbZhMCKJwHzCa9. If this adds complexity we can organise it here or if someone has Calendly they could use that.

**Comment by aothms** [comment]
Tue 14 10am GMT+02:00 / 6PM Sydney? @Moult would that work? Earliest slot I saw.

**Comment by theoryshaw** [comment]
Just a heads up.  We'll need to push the current funds ($1,852.46 USD) back to the general OSArch pool after November 13, 2023--As there hasn't been any takers on this issue for a year, at that point. 
https://opencollective.com/osarch/projects/visualize_layers_in_blenderbim



> _"Please note, if no one provides a solution to the project with a year (365days), the funds dedicated to this project will be transferred over to OSarch's general funding. The OSArch steering committee will then allocate the funding to other projects."_

**Comment by arossti** [comment]
Nooooooo! This is so important, let me see if I can recruit an interested party :) How did this end up going cold?

**Comment by aothms** [comment]
We spoke with @rhughes42 but I think what was unhelpful from our side in the end was that we didn't provide a sense of direction, of either:

- implement in occt / slice actual solids, improve perf of existing code we have for this, and more API as opposed to upfront conversion setting
- implement as bpy/bmesh edge subdivide + subsequent extrusion calls
- implement as a purely visual overlay

I think option 2 is the most feasible. Because anything C++ is just so much infrastructure to take care of. It also allows snapping. As opposed to overlay we would somehow also need to implement the reverse though to toggle back and forth between layered and non-layered display mode.

Then we should set out a quick game plan. I did that, but I think failed to send it to @rhughes42 and now cannot find it. But roughly it would be:

1. identify longitudinaal face styles based on first and last layer in set
2. subdivide the edges in thickness direction based on layerset build up
3. extend / contract generated side faces in simple end-to-end orthogonal case, taking into account priorities
4. work towards general case of arbitrary angles and connections "atpath"

Maybe you can take a leading role in overseeing this @theoryshaw with some visuals of the 4 steps and some test cases. If you agree to this approach that is....

**Comment by arossti** [comment]
fwiw, this is how ArchiCad does it: https://help.graphisoft.com/AC/27/INT/_AC27_Help/025_Attributes/025_Attributes-23.htm, but there are too many over-rides in other parts of the software that make it complicated for users, namely, layer-based priorities (layer as in CAD layers, ie. 'Interior Walls' can have a lower priority than 'Exterior Walls'), and over-rides at the element level (ie. a specific column can punch through 20 stories of floor slabs). It is FAR better if priorities are set in only one place - driven by the materials, as shown in this GS help uri. If you need two flavours of intersection priority for concrete, make 2 types of concrete, easy to manage.

**Comment by Moult** [comment]
I agree with option 2. It's more about figuring out strategies that work reliably at the moment, rather than having to use a particular kernel or language.

If this were implemented purely in Blender for now, I think one strategy that works is to take the existing code that handles join_L and join_T, and refactor it so that instead of looping once per wall junction, it loops once per wall layer per wall junction. Then, the results are unioned together. There is a bit of trickiness to handle join_T (ATPATH) connection priorities, but probably solvable.

I'm going to take a leap of faith and take this on this release cycle and see what happens.

**Comment by arossti** [comment]
Amazing - thank you. Sent a tip to OC :) Hopefully others can match!

**Comment by rhughes42** [comment]
Hey guys, good to see that the discussion around this is still going. If you have those files or examples ready I would like to look at them out of curiosity, that is if someone else is working on it, or if not, to look into it again :)!

**Comment by arossti** [comment]
Just a tangentially related question, when creating materials (to be used in layers/composite assemblies) for the purpose of wbLCA, GPT suggested creating PSets for LCA properties under this: https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcClassificationReference.htm - for example when we are referencing ISO 21930:2017, typical properties for 'Douglas Fir Lumber' might be structured like: 

`/* ------------------------------------------------------------------------------
 * ISO 21930:2017 Environmental Product Declaration for Douglas Fir Lumber
 * IFC 4.3 Property Set Definitions for Material Environmental Impact Indicators
 * ------------------------------------------------------------------------------
 */

/* Material Definition */
# IfcMaterial:
   Name: "Douglas Fir Lumber"
   Description: "Timber product used in structural applications, sustainably sourced."

/* Material Properties (Relates to lifecycle stages as per ISO 21930:2017) */
# IfcMaterialProperties:
   # Pset_MaterialEnvironmentalImpactIndicators:
      {
         GlobalWarmingPotential: {
            A1: 50 kg CO₂ eq,
            A2: 5 kg CO₂ eq,
            A3: 3 kg CO₂ eq,
            A4: 1 kg CO₂ eq,
            A5: 2 kg CO₂ eq,
            B1: 0 kg CO₂ eq,
            B2: 0.5 kg CO₂ eq,
            B3: 0.2 kg CO₂ eq,
            C1: 0.8 kg CO₂ eq,
            C2: 1 kg CO₂ eq,
            C3: 2 kg CO₂ eq,
            C4: 1 kg CO₂ eq,
            D: -5 kg CO₂ eq
         },

         AcidificationPotential: {
            A1: 0.3 kg SO₂ eq,
            A2: 0.05 kg SO₂ eq,
            A3: 0.02 kg SO₂ eq,
            A4: 0.01 kg SO₂ eq,
            A5: 0.03 kg SO₂ eq,
            B1: 0 kg SO₂ eq,
            B2: 0.01 kg SO₂ eq,
            C1: 0.02 kg SO₂ eq,
            C2: 0.03 kg SO₂ eq,
            D: -0.01 kg SO₂ eq
         },

         EutrophicationPotential: {
            A1: 0.01 kg PO₄ eq,
            A2: 0.005 kg PO₄ eq,
            A3: 0.002 kg PO₄ eq,
            A4: 0.001 kg PO₄ eq,
            A5: 0.003 kg PO₄ eq,
            B1: 0 kg PO₄ eq,
            B2: 0.0005 kg PO₄ eq,
            C1: 0.001 kg PO₄ eq,
            C2: 0.002 kg PO₄ eq,
            D: -0.001 kg PO₄ eq
         },

         OzoneDepletionPotential: {
            A1: 0.000001 kg CFC-11 eq,
            A2: 0.0000001 kg CFC-11 eq,
            A3: 0.00000005 kg CFC-11 eq,
            A4: 0.00000001 kg CFC-11 eq,
            A5: 0.00000002 kg CFC-11 eq,
            B1: 0 kg CFC-11 eq,
            B2: 0.00000001 kg CFC-11 eq,
            C1: 0.00000001 kg CFC-11 eq,
            D: -0.00000001 kg CFC-11 eq
         },

         PhotochemicalOzoneCreationPotential: {
            A1: 0.1 kg C₂H₄ eq,
            A2: 0.05 kg C₂H₄ eq,
            A3: 0.02 kg C₂H₄ eq,
            A4: 0.01 kg C₂H₄ eq,
            A5: 0.03 kg C₂H₄ eq,
            B1: 0 kg C₂H₄ eq,
            B2: 0.01 kg C₂H₄ eq,
            C1: 0.02 kg C₂H₄ eq,
            D: -0.01 kg C₂H₄ eq
         },

         AbioticDepletionPotential_Fossil: {
            A1: 120 MJ,
            A2: 25 MJ,
            A3: 30 MJ,
            A4: 10 MJ,
            A5: 20 MJ,
            B1: 0 MJ,
            B2: 5 MJ,
            C1: 8 MJ,
            D: -15 MJ
         },

         PrimaryEnergyDemand_Renewable: {
            A1: 150 MJ,
            A2: 10 MJ,
            A3: 20 MJ,
            A4: 5 MJ,
            A5: 15 MJ,
            B1: 0 MJ,
            B2: 3 MJ,
            C1: 7 MJ,
            D: -10 MJ
         },

         PrimaryEnergyDemand_NonRenewable: {
            A1: 80 MJ,
            A2: 15 MJ,
            A3: 25 MJ,
            A4: 10 MJ,
            A5: 5 MJ,
            B1: 0 MJ,
            B2: 2 MJ,
            C1: 5 MJ,
            D: -5 MJ
         },

         WaterConsumption: {
            A1: 2 m³,
            A2: 0.5 m³,
            A3: 1 m³,
            A4: 0.2 m³,
            A5: 0.3 m³,
            B1: 0 m³,
            B2: 0.1 m³,
            C1: 0.4 m³,
            D: -0.2 m³
         },

         WasteHazardous: {
            A1: 0.5 kg,
            A2: 0.1 kg,
            A3: 0.2 kg,
            C4: 0.4 kg
         },

         WasteNonHazardous: {
            A1: 5 kg,
            A2: 1 kg,
            A3: 2 kg,
            C4: 4 kg
         },

         WasteRadioactive: {
            A1: 0.001 kg,
            A2: 0.0005 kg,
            A3: 0.0002 kg,
            C4: 0.001 kg
         }
      }

/* Classification References (for ISO 21930:2017) */
# IfcClassificationReference:
   ClassificationName: "ISO 21930:2017"
   Edition: "2017"
   Identification: "Environmental Product Declarations for construction products"
   Source: "ISO"
`

**Comment by theoryshaw** [comment]
For what it's worth, i started a mockup.  Hope to share in a week or so.

**Comment by Moult** [comment]
Now that item editing is starting to settle, it's time for me to focus on this. Here's my attempt at itemising what I think needs to get done to implement it (it might seem a bit roundabout, but there's a logic and precedence to the items I'm hoping for).

- [x] Fix existing bugs around profile editing, particularly around the assignment of arcs, and consecutive arcs. This is to ensure that our profile editing strategy is robust enough to handle wall axis editing later on.
- [ ] Implement 2-point arcs in the profile editor. Currently this only exists in the regular mesh CAD tool, but it's likely to be one of the most common ways arcs are actually drawn for walls.
- [ ] Allow tabbing into LAYER2 elements and either generating or fetching the axis representation for editing.
- [x] Adjust the profile tool to support open profiles. This will be useful later on for editing the wall axes, and even in advanced cases allowing us to handle the `CurveOnRelatedElement` geometry which is later necessary for parametric wall joints.
- [ ] Implement walls which have arcs as their axis (or any arbitrary curve, but starting with a UI for indexed polylines only).
- [ ] Deprecate butt and mitre joints in favour of material layer priority.
- [ ] Refactor the wall intersection algorithm to be agnostic of a wall element, but instead operate with pure geometry. Instead of outputting wall parameters and clipping planes, it should instead output end-cap curves (i.e. `CurveOnrelatedElement) and some sort of unioned footprint, or unioned boolean geometry. The boolean geometry can no longer be as simple as a half space solid due to the possible concave intersection curves.
- [x] Refactor the wall junction to loop through layers and run the wall intersection algorithm per layer.
- [ ] Come up with a modified wall intersection algorithm for straight + arc, or arc + arc walls. This will operate on the last segment of an indexed polyline to allow walls to extend to one another.
- [ ] Develop a headache around arcs and sloped walls. At this point, we should now have correct body envelopes of walls, but still (!!!) no layer visualisation yet.
- [ ] Develop a 2D only layer slice visualisation for sections, probably using shapely intersections of the cut line and projected offsets of the axis curve.
- [ ] Actually get each layer polygon to show up with drawings with semantic material fills and styling.

**Comment by aothms** [comment]
List looks good to me, but it would help me understand if we specifically map out the interaction points between C++ and python.


> The boolean geometry can no longer be as simple as a half space solid due to the possible concave intersection curves.

Technically speaking a halfspace solid does not necessarily need to conform to a plane: https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcHalfSpaceSolid.htm Not sure though how many implementations support non-planar halfspaces

**Comment by Moult** [comment]
At the moment I'm not sure what the issues will be in IfcOpenShell. I'm breaking down the problem into three main areas:

 1. Bonsai UX, so that we can actually stress tests the results later
 2. Generating the correct body of the wall, with the correct footprint and possible booleans at start/end/path
 3. Slicing the body of the wall into layers, this is the bit which you've already got some logic in C++ for I believe. But I think there's quite a lot of ground work in 1. and 2. before I get to this stage.

> a halfspace solid does not necessarily need to conform to a plane

... whoah.

I don't think it'll work still as a general solution, because it doesn't work for ATPATH connections where the joining wall has a higher priority (i.e. the joining wall may actuallly split the first wall into two pieces).

**Comment by theoryshaw** [comment]
> For what it's worth, i started a mockup. Hope to share in a week or so.

Or in 3 weeks. :)

I took a stab at creating a mockup...
https://hub.openingdesign.com/OpeningDesign/Visualize_IfcMaterialLayerSets
_(clone recursively, so the git submodules get pulled down as well)_

There's a few brainstorming proposals in there, as well as some assumptions for future functionality, such as
-  'joining' things
- being able to 'butt/miter' things like AXIS3 objects
- extending (1) AXIS2 objects to multiple AXIS3 objects. 

The brainstorming ideas...
- If layer priorities are equal, joining order takes precedence. 
- If a layer is 'cutoff' by a layer with a higher priority, that layer does not carry through.

Not sure how doable these are, relative to the schema, but thought i'd just put them out there for discussion. 

...

There's also a few modeling/drawing glitches called out in the model, as well.



![test](https://github.com/user-attachments/assets/dcad82fe-ee05-44ef-ba48-6589c66caeb4)





https://github.com/user-attachments/assets/9b6342a6-6568-431f-abba-0aaca66ab3b3

**Comment by arossti** [comment]
I wouldn't want to start posting details that aren't directly related to the specific issues of BIM intersection priorities, etc. but we have moved from a style of BIM detailing that accounted for every layer of an assembly with a thickness (ie. yes even providing a thickness to membrane layers which distorted real dimensions but provided better graphic display). We have since reverted to line-type conventions between materials to account for such barriers. What do I mean? One image: 
<img width="818" alt="Screenshot 2024-10-12 at 15 17 51" src="https://github.com/user-attachments/assets/f42a7b09-b8eb-48ac-80af-002f6948de05">
This list of materials by thickness is generated automatically. It reads the thickness and the material. I could just show the membranes with a graphic linetype, but then I would lose all automation for labelling, and some contractor might even omit a membrane layer because it gets missed in our notes. By modelling even thin stuff like flashings and membranes, they get included in our specs, lists, BOQ and everything else - AND they can participate in priorities. For example, a good air barrier layer should not be able to be penetrated by almost anything, as it plays a vital role in building performance. And yes on this job there were two drywall layers with a membrane (AVB) sandwiched in-between, each showing their own fill. The Type-X gypsum and the SIGA Majvest layers have super high priorities because their continuity is important. So as part of this thread, our aim is to develop two templates, two workflows, one for what we can part 9 residential construction (stick frame details) and one for commercial mass-timber or concrete shearwall construction. Rockwool has produced two really good guides full of details that represent how we would want drawings to look - but also demonstrate that this is not how we model things in BIM, where layers kind of require to touch each other and aren't separated for graphic clarity. The guides, that will closely accord with our templates are available online for free here: 

Residential Guide: https://p-cdn.rockwool.com/siteassets/o2-rockwool/documentation/construction-details/residential/constructiondetails-woodframe-construction-lightweight-horizontal-cladding.pdf?f=20210414142440
Commercial Guide: https://www.rockwool.com/siteassets/o2-rockwool/documentation/technical-guides/commercial/passive-house-solutions-guide.pdf

Ultimately I'd love to generate all of the profiles for use in Bonsai for flashings, and ask for some help from Bruno P to script these details to generate from topologic cell complexes per his Homemaker add-on logic. If the materials and their graphic representation can get nicely sorted in IFC, then generating complete details for entire buildings could be automagically completed with Homemaker, right down to BOQ, Cost and wbLCA analysis and integration with specifications. PS - with those guides posted, you could easily sub-in EPS, GPS, TimberMX or any other board-product insulation...

**Comment by arossti** [comment]
<img width="970" alt="Screenshot 2024-10-12 at 15 45 17" src="https://github.com/user-attachments/assets/e082d853-cb24-4cc9-811b-ab8137855ef6"> 95% of these details are 3D BIM (except dims/notes which in AC are 2D). This whole building was built from basically one polyline that generated most of the assemblies.

**Comment by theoryshaw** [comment]
It would seem creating layers only in the SVG, would be a nice intermediate step.

**Comment by MarcelP102** [comment]
Any update?

**Comment by Moult** [comment]
![Image](https://github.com/user-attachments/assets/686223fd-093a-4993-802e-0dfb11be3eb7)

**Comment by Moult** [comment]
OK there is now a standalone test script that can produce the results above. A few things I like:

 - Priorities work! This will make the existing mitre/butt features obsolete.
 - Joint-specific priority overrides are supported too (something like Revit's wall join tool)
 - All geometry is done using a profiled swept solid extrusion. No booleans (fast!)
 - All computation is done in local coordinates axis aligned to the +X axis. This means there isn't any code which does angle calculations. Hopefully this is more robust.
 - All calcs done in project units.
 - All math done with numpy directly with IFC. No Blender required.
 - All math done in 2D.
 - Direction sense, reference line offsets, axis context lines supported, atstart/atend. Mostly feature parity.

Some known limitations:

 - Will not work on sloped walls. Although I've been really tossing up between generating complex 3D booleans, I'm currently leaning towards (get it!?) just repeating the 2D analysis and then generating a mesh. This means we can also support wall junctions on two walls which have different heights.
 - Not yet generating the actual layer division lines. Code functionality is there, just needs a bit of reorganisation.
 - In Revit, it doesn't quite work like IFC, but it has an interesting feature where you can toggle "wrap" of non-core layers where the layer wraps around a butt joint. IFC doesn't have this, but ... how do I explain it, you win some, you lose some.
 - No curved wall support. I plan to just ignore it for now and add it later. The foundation will be reusable though.
 - Still work to do with changing object placements
 - Haven't yet written code that checks for impossible connections (parallel, flipped ends, negative length walls...)
 - No support yet for "sliding" openings (i.e. openings not moving based on their previous spot in the local X axis).
 - No support yet for ATPATH junctions.
 - No support yet for custom connection geometry.

![Image](https://github.com/user-attachments/assets/5158d20e-aa1c-4cd6-8240-e0a3273883f0)

**Comment by aothms** [comment]
> In Revit, it doesn't quite work like IFC, but it has an interesting feature where you can toggle "wrap" of non-core layers where the layer wraps around a butt joint

Is that the 'fold layerset' option we have in ifcconvert?

![Image](https://github.com/user-attachments/assets/925fff16-b6c5-4865-9119-6dccb5a7403d)

**Comment by theoryshaw** [comment]
Awesomesauce. 

Dumb question, will selection order come into play? 
If not, could it?

![Image](https://github.com/user-attachments/assets/4e4be813-665f-443b-bc03-623b1ca23244)


<details>
<summary></summary>

https://excalidraw.com/#room=47cfe8ca06dd7f820253,F4-N5xkodwUY4MaxwtUNGQ

</details>

**Comment by RayAtGit** [comment]
In Revit layered constructions can be duplicated as "layered" construction. By selecting the correct view option you see the layered, the original or both.

When the view "layered" is activated you can activate shape handle  on the layers and adapt them (not to confuse with the corner joint function in Revit) 

The layers can be edited by slicing and joining them. Giving a high degree on functionality also for corners and joints.

Real life example., the area where an prefabricated element wall is lifted up in the area where the stairs are.

3D:
![Image](https://github.com/user-attachments/assets/f8dc9e19-2e70-4ead-b023-ffc9351f9be4)^

3D non layered view (the original wall) 
![Image](https://github.com/user-attachments/assets/cf11aac4-a94a-443e-bbea-e75255270345)

2D:
![Image](https://github.com/user-attachments/assets/b537003b-dcb7-427e-972c-f5558f2b50cf)
Sorry for the colored distance lines... we are in the middle of a review process

Cut: with "activated shape handle"
![Image](https://github.com/user-attachments/assets/43709ef4-7d10-40fb-ba1d-47e1482a09c7)

So IMHO the functionality to manipulate layers should be an integrated part of the any joint solution (corners and so on).

The layers are not only for walls but for any construction with multiple layers like floors!

**Comment by Moult** [comment]
> Is that the 'fold layerset' option we have in ifcconvert?

Yep

> will selection order come into play?

No. Priorities determine everything.

> If not, could it?

Yes, after the joint, we can build a "Wall Join Tool" feature which overrides the join, which takes into affect selection order. This is a later priority.

> In Revit layered constructions can be ...

In our first pass, we are implementing layers exactly to the IFC specification. Only after we've done this, will we look at ways to go further.

> When the view "layered" is activated you can activate shape handle on the layers and adapt them

In IFC, this is known as the "connection geometry" feature. It only works in the XY plane, so layers protruding up in the Z direction will need to be done in a different way, most likely with another covering element or boolean.

I forgot to put it in the "known limitations" section, I'll add it.

**Comment by RayAtGit** [comment]
> In IFC, this is known as the "connection geometry" feature. It only works in the XY plane, so layers protruding up in the Z direction will need to be done in a different way, most likely with another covering element or boolean.

For the connection geometry this is correct ... But please keep in mind that the connection geometry probably wants to follow the logic of an ifcmaterialsetlayer.

https://standards.buildingsmart.org/IFC/DEV/IFC4_2/FINAL/HTML/schema/ifcmaterialresource/lexical/ifcmateriallayersetusage.htm .... Generally, an element may be layered in any of its primary directions, denoted by its x, y or z axis. 

Examples for Wall and Slab are depicted a bit further in the description.

**Comment by Moult** [comment]
@RayAtGit yep, the existing wall code already fully implements IfcMaterialLayerSet (and more specifically, its applicability to IfcWall(StandardCase). This issue is about visualisation, but a precursor to visualisation is first finishing the implementation of IfcRelConnectsPathElements (previously only half implemented), which is what you're seeing in the above screenshots.

3D viz is looking good now to me. Colours are based on layer material styles. Snapping is supported.

![Image](https://github.com/user-attachments/assets/22028b70-ba7a-4a36-9df8-e563d1940ec3)

**Comment by Moult** [comment]
So doing ATPATH joins was a little bit more involved. Basically if you want to avoid booleans (hopefully this is the right way and worth it in the end) you have to generate the footprint profile. But a wall can have all sorts of notches cut into it (or all the way through, which effectively splits the wall!) due to ATPATH joins from either end. It's also possible to have multiple intersecting ATPATH junctions that are basically just invalid and unconstructible in real life (which probably means you should model your object with a bit more care).

![Image](https://github.com/user-attachments/assets/72b9a42f-b896-4fe9-95ad-82141527f61f)

Anyway it seems to behave now. Probably still needs more defensive coding somewhere.

![Image](https://github.com/user-attachments/assets/48e29c4f-ee34-45e1-a25c-c76686607285)

Next up is sloping walls which is probably another kettle of fish and the make or break on whether or not avoiding booleans was worth it or not.

**Comment by aothms** [comment]
I wonder if it makes sense to constrain this, that if you have an atpath connection one is not allowed to have priorities that split the other wall? In the spec or in the tool...

**Comment by Moult** [comment]
Sweet! Sloped joins work!

If no slopes are detected, it merely generates a profiled extrusion (which may be a composite profile, if the wall is split by an ATPATH connection). If an angled connection is detected, it will then instead create booleans and do any potential expansion of the base profile so that it can be booleaned. Right now the booleans are exact, i.e. no dilation of the boolean for tolerance. IfcOpenShell seems to deal with it pretty well.

The shrinking list of known limitations / TODO is:

- Need to actually integrate this code into Bonsai itself. That'll also help in stress testing as it's easier to draw wacky shapes.
- No curved wall support. I plan to just ignore it for now and add it later. The foundation will be reusable though.
- Still work to do with changing object placements
- Haven't yet written code that checks for impossible connections (parallel, flipped ends, negative length walls...)
- No support yet for "sliding" openings (i.e. openings not moving based on their previous spot in the local X axis).
- No support yet for custom connection geometry.
- Visualisation of layers in Blender relies on a mesh based plane bisection. Due to Blender's mesh tolerance sometimes if you look inside a wall join the wrong colour will be shown. I'm treating this as a non-issue for now.

![Image](https://github.com/user-attachments/assets/8ef5dba5-e248-4c0b-b569-c523e8774367)

**Comment by brunopostle** [comment]
Cool, note that there probably isn't a general solution for mitreing curved walls, the mitre plane is a complex curve:

![Image](https://github.com/user-attachments/assets/b1b92ca1-e7af-4c3e-87ba-ddf155fb7adf)

**Comment by theoryshaw** [comment]
The slices don't seem to be rendering correctly. 
Most of these walls should show drywall, not brick. 

file: https://hub.openingdesign.com/OpeningDesign/OD_Project_Template/src/branch/main/Open/Models/Bonsai

https://github.com/user-attachments/assets/d6db3dfd-0616-4ece-a889-65a6673f4626

**Comment by theoryshaw** [comment]
you probably know this, but after enabling layerset slicing, the slabs are sliced in the wrong direction.

![Image](https://github.com/user-attachments/assets/4cb9c805-e6f7-4181-aab8-2a06a38126d4)

**Comment by theoryshaw** [comment]
Would it be possible to have each layer be a watertight cube?  Currently it appears the layers are only shown at the edge. 
The video clip below illustrates.

If the layers are watertight cubes, they can be 'exploded' later in the design process to make independent geometry, that quantity analysis can be run on. 

They also would hold up better in renderings, when one of the layers is 'air'. 



https://github.com/user-attachments/assets/978f4fe7-9f7b-4b4c-9b55-a199a3eb2f6a

**Comment by Moult** [comment]
Yes slicing into manifold solids is possible, but let's leave that on the todo list. Here's my updated todo list:

- Need to actually integrate this code into Bonsai itself. That'll also help in stress testing as it's easier to draw wacky shapes.
- Wall join and unjoin
- Wall split and merge
- Wall extend
- Wall flip
- Check top/bottom of slabs I think it's wrong currently
- Still work to do with changing object placements
- Haven't yet written code that checks for impossible connections (parallel, flipped ends, negative length walls...)
- No support yet for "sliding" openings (i.e. openings not moving based on their previous spot in the local X axis).

TODO later:

- See layered visualisation in drawings
- No curved wall support
- No support yet for custom connection geometry.
- Manifold solid slicing

Known limitation that I don't plan to address:

- Visualisation of layers in Blender relies on a mesh based plane bisection. Due to Blender's mesh tolerance sometimes if you look inside a wall join the wrong colour will be shown. I'm treating this as a non-issue for now.

**Comment by Moult** [comment]
OK it's a bit more ready for stress testing within Bonsai. Let me know if you spot any crazy behaviour. Revised TODO list:

- Check top/bottom of slabs I think it's wrong currently
- See layered visualisation in drawings

TODO later:

- No curved wall support
- Manifold solid slicing
- Wall connection priority override tool
- Custom connection geometry tool
- Reimplement auto slab clipping (right now it'll be purged)
- Reimplement manual booleans (right now it'll be purged)

Known limitation that I don't plan to address:

- Visualisation of layers in Blender relies on a mesh based plane bisection. Due to Blender's mesh tolerance sometimes if you look inside a wall join the wrong colour will be shown. I'm treating this as a non-issue for now.

**Comment by brunopostle** [comment]
Layer slicing! Here's a corner case that doesn't look right, the slab is extruded non-orthogonally and the layers follow the extrusion direction rather than being parallel to the profile:

![Image](https://github.com/user-attachments/assets/195527a7-6c0d-4477-8eb0-e3fd582532fe)

[slab-representation.ifc.txt](https://github.com/user-attachments/files/19152621/slab-representation.ifc.txt)

**Comment by sboddy** [comment]
@Moult Regarding

> - See layered visualisation in drawings

Can I make a small request: That the layering in 2D drawings be optional. For planning permission (UK) it is not necessary to show plans with full layering. Thanks!

**Comment by Moult** [comment]
@sboddy yep :)

So with the latest commits the visualisation of layers in the viewport when cutting is mostly there (the old code has been removed and wasn't quite correct anyway). There are some problems still with bad cuts due to tolerance so it needs a bit more work still. But the good news is that I can see the light at the end of the tunnel. Revised todo:

 - Fix reordering of layers regression for negative sense slabs
 - Tolerances fixes on mesh bisections
 - See layered visualisation in SVG

Todo later:

 - No curved wall support
 - Manifold solid slicing
 - Wall connection priority override tool
 - Custom connection geometry tool

Known limitation that I don't plan to address:

 - Visualisation of layers in Blender relies on a mesh based plane bisection. Due to Blender's mesh tolerance sometimes if you look inside a wall join the wrong colour will be shown. I'm treating this as a non-issue for now.

Also drawings now clip. Sweet.

![Image](https://github.com/user-attachments/assets/8468d39e-2424-49f7-92fd-526865ec6962)

![Image](https://github.com/user-attachments/assets/0cd907d8-af02-413b-86f5-d0d847642332)

![Image](https://github.com/user-attachments/assets/4834abe4-1aec-4481-8612-77b3b2fcc635)

**Comment by theoryshaw** [comment]
this might be known already, but with `shift +e` it extends to the top of the slab.

file: https://hub.openingdesign.com/OpeningDesign/Visualize_IfcMaterialLayerSets/src/commit/928762e3f51221a871197b871fe1b0d032bb0cb0/Visualize_IfcMaterialLayerSets.ifc

![Image](https://github.com/user-attachments/assets/aeb90a45-4204-402b-80f2-06954ef25add)

**Comment by Moult** [comment]
The "extend to slab" feature is now more accurately "extend to underside". I.e. it finds faces with normals pointing down of the object that you are extending to, and uses that as a clipping. I just fixed a few incorrect normals which would lead to the experience you have.

Another thing to be aware of is that in the new implementation, it (on purpose) does _not_ regenerate. It's a baked clipping geometry. That means that if you then move the roof or the wall, the old clipping geometry will still be there and you need to manually remove it and add a new one (by doing Shift-E again).

Here's a video of me deleting your old clipping geometries, and reclipping it. Notice how it's not perfect, sometimes the clip won't quite work, or IOS might choke, but just going in and manually cleaning the clipping will fix it. It should work in most cases though.

https://github.com/user-attachments/assets/60776e0f-a4d0-4f79-8617-ba456d087667

**Comment by Moult** [comment]
OK we've got something. Fingers crossed it's somewhat reliable to use Blender mesh "good-enough" bisection for this coupled together with shapely which is super strict. There is a new layer-material-foo class you can use to style with CSS.

Needs a ton of testing, and a bit more polish as well as the ability to toggle on/off, but I'd say this is the beginning of the end of this bug. The rest (i.e. curved walls, new priority UX tools) can come later.

![Image](https://github.com/user-attachments/assets/cd57d461-23df-4052-9f11-13e0b848462f)
![Image](https://github.com/user-attachments/assets/bd07d153-de7d-4d2c-80ed-7b6ca9d30cc4)

**Comment by theoryshaw** [comment]
awesome.

Is there any way to put something like `layer-material.Category` to style layers by the material's category?

**Comment by SavuGust27** [comment]
Not sure if intended or a bug, but the decorators persist when activating view in another story on some walls

![Image](https://github.com/user-attachments/assets/270e6c44-ee18-489f-9ad4-c50de689acfe)

![Image](https://github.com/user-attachments/assets/32b94a18-1f8f-4960-9506-5efd69d170bf)

**Comment by Moult** [comment]
I couldn't replicate it exactly but I found something similar and I think it's the same issue.

**Comment by theoryshaw** [comment]
Closed! Thanks to all! 
https://www.linkedin.com/feed/update/urn:li:activity:7309976615596302339/

**Comment by theoryshaw** [comment]
> Another thing to be aware of is that in the new implementation, it (on purpose) does not regenerate. It's a baked clipping geometry. That means that if you then move the roof or the wall, the old clipping geometry will still be there and you need to manually remove it and add a new one (by doing Shift-E again).

@Moult, 
Any objections to making this automated again? 
Is there a use case you had in mind that is more amenable to not have this regenerate automatically?
In my view, being able to quickly regen the wall after moving a roof up/down (for example) is a great time saver during design.
